Add Support for Configuring Slice View Annotations

arhowe00 opened the pull request:
## Summary

This Pull Request attempts to address [Refactor Slice View Annotations infrastructure #7184](https://github.com/Slicer/Slicer/issues/7184) by separating the existing annotations infrustructure into components that support flexibilty. The intention is also to rely less on Python wrappers and bring command/observation routines from the the interpreted layer into C++. The functionality is broken down into three main parts:

1. A prescriptive, XML list of tags desired in each corner, currently a `vtkMRMLTextNode`:
  ```XML
  <annotations>
    <corner position="bottom-left">
      <property name="Label" prefix="L: " display-level="always"/>
      <property name="Foreground" prefix="F: " display-level="always"/>
      <property name="Background" prefix="B: " display-level="always"/>
    </corner>
    <edge position="bottom">
      <property name="Label" prefix="L: " display-level="always"/>
    </edge>
  </annotations>
  ```
2. A public API allows plugins to substitute the XML with annotations, given the `<property>` tag's attributes:
  ```C++
  virtual bool
  CanProvideValueForPropertyName(const std::string &propertyName) = 0;
  virtual std::string GetValueForPropertyName(const std::string &propertyName,
                                          const XMLTagAttributes &attributes,
                                          vtkMRMLSliceNode *) = 0;
  virtual std::unordered_set<std::string> GetSupportedProperties() = 0;
  ```
3. Application logic for querying the registered APIs & XML and separate rendering routines, implemented in C++ (see diffs for specifics).

---

## In reference to [#7184](https://github.com/Slicer/Slicer/issues/7184)

This solution accommodates @jcfr's requested solution in the following ways:

- Re-implemented in C++
> The solution is implemented in C++.
- Organize the content to display in `vtkMRMLTextNode` nodes with associated displayable manager
> Each slice node associates with a `vtkMRMLTextNode` where the annotations are prescribed through the simple XML. By default, all slice nodes associate with the same one, as is the current functionality of `SliceViewAnnotations.py`. However, the current form makes it easy for both Layer- and SliceNode specific annotations to be provided
- Allow customization of displayed fields
> Can be handled via the provider & XML
- Support displaying
  - Slice position
  - Anatomical direction codes
  - Image display parameters (window, level, slice thickness, etc)
  - Patient/study/series metadata
> Can be handled by the provider & XML (examples are below of doing this for DICOM data). In sum, a provider knows the specific slice node and any number of XML tag attributes, and the provider can decide what should be displayed specific to a given slice node. Since annotations are rendered on the slice view, we assume that passing the MRML slice node to the provider should be everything that is needed to know what should be displayed.

---

## Commentary

For now, we have not added boolean flags to each corner of `vtkMRMLSliceNode`s for displaying annotations selectively in slice nodes. However, it would be simple to implement this given the modular framework of the refactoring. Each slice node could have boolean flags that are checked by the Displayable Manager to render specific corners on specific slices, something not available with DataProbe's `SliceViewAnnotations.py`.

The [first commit](https://github.com/Slicer/Slicer/commit/e2bbacd44e003ebb64fdc2914f296a0854fdb8b2) contains changes to MRML (CornerText logic and Abstract API), adds a Loadable Module, implements the API in a default case to mimic DataProbe's `SliceViewAnnotations.py`, and removes `SliceViewAnnotations.py` to prevent clashes. To understand why we added anything into MRML core at all, see the Q&A below. The default example uses the API to register a 'Default' annotation provider, which displays the volume names in the bottom-left corner, *but not anything related to DICOM*. You can verify this works by checking out, building, and loading data. For instance:

![image](https://github.com/user-attachments/assets/7f586400-9b1a-4c18-9491-72fdaf2a39c5)

You can also see that by editing the text node directly, within the "Texts" module, you can change the annotations using the prescriptive XML. While we like this approach, we are open to a programmatic one, see the commentary below for more information and discussion. 

![image](https://github.com/user-attachments/assets/dd2def35-4081-4e7f-85cc-451927e919dd)

Noticeably, we did not register a DICOM provider in the first commit. Since this is important, but we were unsure whether this XML + Provider API construct would survive, we made a "high-level" DICOM value provider that intends to exactly emulate the behavior of DataProbe (with some edge-case differences that can also be seen as bug fixes). It is in the [second commit](https://github.com/Slicer/Slicer/commit/e2bbacd44e003ebb64fdc2914f296a0854fdb8b2). As you can see, adding DICOM as a provider shows the extensibility of this approach. The database is queried in Python, and the C++ wraps those calls to provide `vtkMRMLCornerTextLogic` with the values expected in the XML tags. Since the registration process is separate from the rendering routines and logic, the way DICOM tags are provided in `DICOMAnnotationPropertyValueProvider.py` is much more intuitive than in `SliceViewAnnotations.py`. Here is a side-by-side depiction of 3D Slicer 5.6.2 and the feature branch with DICOM loaded in both layers.  

![image](https://github.com/user-attachments/assets/a595972a-1bb7-4eec-a788-75255b2bff77)

Should our proposed API be favored, we will allow tag providers to be registered directly in Python so that the DICOM database calls and caching of results can be implemented in a single Python file.

---

## Q&A

### Q: Why did you add or change anything in MRML Core?

A: To avoid introducing a dependency from the core to a module, the `vtkSlicerCornerTextLogic` class was moved from being in a Loadable module to Libs/MRML/Logic. Some of the dependencies include `vtkMRMLLayoutLogic::UpdateViewCornerAnnotations()` and `vtkMRMLApplicationLogic::GetCornerTextLogic()`. We suggest associating a single `vtkMRMLCornerTextLogic` with the application logic.

---

### Q: Why prescriptive XML?

A: The prescriptive XML, at least currently, allows for high-level and simple modularity available both to developers and users. While a programmatic approach is needed to know what a tag should finally print on the slice view, a user can easily change, e.g., the corner that text is in. We think the XML logic we define is simple enough that all users should be able to manipulate the display of provided tags as desired. However, parsing XML is an inflexibility and could introduce challenges for competing plugins.

---

### Q: Why add an API and not use MRML alone?

A: There is an ongoing discussion of how we can minimize and simplify all exposed functionality. While we think the use case for the API will be isolated enough that developers can efficiently invest their time into implementing the API for their extension, @jcfr and I have been discussing how we can directly address node attributes and make a more MRML-native way to define corner annotation values. However, whatever way we choose, the infrastructure for returning tag values must be implemented somewhere. In the case of this PR, we have decided to do it in a separate API, but we may scrap this idea.

---

### Q: What will be done for modules that want to append annotations as opposed to monolithically overwrite a `vtkMRMLTextNode`?

A: We understand the fragility of the current approach and have thought of a few options. We could make it so that providers also register their own XML (assuming we continue with XML) and then have the `vtkMRMLCornerTextLogic` process all provided XMLs into a final annotation `vtkMRMLTextNode` that is displayed in the current manner.  In this case, We could give each module registering a provider a name (e.g. `Default`, `DICOM` -- perhaps just the name of the module). Annotations could be rendered in the order of a **floating priority** that would a reserved tag attribute (in the XML case). 

If there are clashes, we could additionally add a "rare-case" function that has an API such as setTagOrder(["MyProvider", "Default", "DICOM"]); that will allow a user to impose a second round of ordering based on the descriptive tag order. To minimize confusion, this should be more hidden in the APIs, as this case should be extremely rare. It's pretty clear what the function does and how it works, should the need arise.

---

### Q: What will happen to modules and dependencies that use `SliceViewAnnotations.py`? 

A: There are some codebases we have found that import classes defined in `SliceViewAnnotations.py`. Since this PR will end up deprecating those classes, we should address the ins and outs of those dependencies and whether we should provide backup functionality.

arhowe00 committed:
ENH: Refactor Slice View Annotations Infrastructure

This commit refactors the Slice View Annotations infrastructure to
enhance the flexibility and maintainability of DataProbe
SliceViewAnnotations.py. By migrating core annotation logic to C++ and
decoupling the rendering from the annotation generation and provider
registration logic, it is much easier for 3rd party extension providers
to render text on slice views. The changes address the requirements
outlined in issue #7184.

1. **Prescriptive XML for Corner Annotations:**
   - Introduced a `vtkMRMLTextNode` containing XML to define annotations
     for each corner and edge of the slice view.
   - Example XML specifies tags, prefixes, and display conditions,
     allowing flexible configurations.

2. **Public API for Annotation Providers:**
   - Defined an abstract C++ interface for plugins to implement custom
     annotation value logic based on tag attributes and slice node
     state.
   - Methods include `CanProvideValueForPropertyName`,
     `GetValueForPropertyName`, and `GetSupportedProperties`.

3. **Centralized Annotation Logic:**
   - Moved annotation querying and rendering routines to C++ for
     improved performance and maintainability.
   - The `vtkMRMLCornerTextLogic` was relocated to MRML Core to avoid
     dependencies on specific modules.

- **DICOM Annotations Not Included:** The DICOM-specific provider is
  part of a subsequent commit to demonstrate the extensibility of the
  new API.
- **Removal of SliceViewAnnotations.py:** To avoid clashes,
  `SliceViewAnnotations.py` was removed in this commit. Dependencies on
  classes defined in `SliceViewAnnotations.py` have not been addressed.

For further details, refer to the discussion and examples provided in PR

Co-authored-by: Jean-Christophe Fillion-Robin <jchris.fillionr@kitware.com>

arhowe00 committed:
ENH: Add DICOM corner annotations provider

Added a DICOM-specific property value provider to the corner annotation
system, enabling the display of DICOM metadata such as patient and study
details. The provider replicates the previous behavior of DataProbe,
using Python for database queries. Currently, it uses a high-level
wrapper to static functions defined in
`DICOMAnnotationPropertyValueProvider.py`, but this should be optimized
if the API logic is not adapted further. The purpose of these changes
are to demonstrate the extensibility of the refactored annotation
infrastructure and ensure compatibility with existing annotation
functionality.

Co-authored-by: Jean-Christophe Fillion-Robin <jchris.fillionr@kitware.com>

jcfr commented on Libs/MRML/Logic/vtkMRMLCornerTextLogic.h:
```diff
@@ -0,0 +1,142 @@
+class vtkMRMLAbstractAnnotationPropertyValueProvider;
+
+// STD includes
+#include <cstdlib>
+#include <array>
```
```suggestion
#include <array>
#include <cstdlib>
```

jcfr commented on Libs/MRML/Logic/vtkMRMLCornerTextLogic.h:
```diff
@@ -0,0 +1,142 @@
+                                               const int viewArrangement,
+                                               const std::string& viewName);
+  bool RegisterPropertyValueProvider(const std::string &pluginName,
+                                vtkMRMLAbstractAnnotationPropertyValueProvider *);
```
```suggestion
                                    vtkMRMLAbstractAnnotationPropertyValueProvider *);
```

jcfr commented on Libs/MRML/Logic/vtkMRMLCornerTextLogic.h:
```diff
@@ -0,0 +1,142 @@
+  vtkTypeMacro(vtkMRMLCornerTextLogic, vtkMRMLAbstractLogic);
+  void PrintSelf(ostream& os, vtkIndent indent) override;
+
+ /// CornerText Logic functions
```
```suggestion
  /// CornerText Logic functions
```

jcfr commented on Libs/MRML/Logic/vtkMRMLCornerTextLogic.h:
```diff
@@ -0,0 +1,142 @@
+  /// Parses the contents of an XMl-based text node into a vtkXMLDataElement.
+  /// The vtkXMLDataElement will be used by other functions to generate the
+  /// actual string intended to be rendered on a given slice view.
+  vtkXMLDataElement* ParseTextNode(vtkMRMLTextNode*);
```
Move to protected

arhowe00 replied:
What did you want me to move to protected here?

arhowe00 replied:
Notes from JC/Andrew 2025.1.20.1157: vtk styling suggests that methods such as `ParseTextNode` 0are suggested to be protected so derived classes can access them.

jcfr commented on Libs/MRML/Logic/vtkMRMLCornerTextLogic.h:
```diff
@@ -0,0 +1,142 @@
+  vtkXMLDataElement* ParseTextNode(vtkMRMLTextNode*);
+
+  bool SliceViewAnnotationsEnabled{true};
+  bool BottomLeftEnabled{true}, TopLeftEnabled{true}, TopRightEnabled{true};
+  int DisplayStrictness, FontSize;
```
Split by having one on each line.

Explicitly initialize `int` as well.

jcfr commented on Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:
```diff
@@ -0,0 +1,358 @@
+//----------------------------------------------------------------------------
+vtkStandardNewMacro(vtkMRMLCornerTextLogic);
+
+//----------------------------------------------------------------------------
+vtkMRMLCornerTextLogic::vtkMRMLCornerTextLogic()
+{
+}
+
+//----------------------------------------------------------------------------
+vtkMRMLCornerTextLogic::~vtkMRMLCornerTextLogic()
+{
+}
```
```suggestion
//------------------------------------------------------------------------------
vtkMRMLCornerTextLogic::vtkMRMLCornerTextLogic() = default;

//------------------------------------------------------------------------------
vtkMRMLCornerTextLogic::~vtkMRMLCornerTextLogic() = default;
```

jcfr commented on Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:
```diff
@@ -0,0 +1,358 @@
+
+//----------------------------------------------------------------------------
+void vtkMRMLCornerTextLogic::PrintSelf(ostream& os, vtkIndent indent)
+{
```
Add display of current properties

jcfr commented on Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:
```diff
@@ -0,0 +1,358 @@
+    return cornerAnnotations;
+  }
+
+  if (textNode->GetText().size() <= 0) return cornerAnnotations;
```
```suggestion
  if (textNode->GetText().size() <= 0)
  {
    return cornerAnnotations;
  }
```

jcfr commented on Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:
```diff
@@ -0,0 +1,358 @@
+      }
+
+      const bool valueNotProvided = propertyName != "" && propertyValue == "";
+
```
```suggestion
```

jcfr commented on Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:
```diff
@@ -0,0 +1,358 @@
+
+    // once each property has been parsed, the annotation for that position
+    // is fully specified
+
```
```suggestion
```

jcfr commented on Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:
```diff
@@ -0,0 +1,358 @@
+      }
+
+      // Get the attribute values out of the XML tag
+
```
```suggestion
```

jcfr commented on Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:
```diff
@@ -0,0 +1,358 @@
+          attributes;
+
+      // Get the value for "name" in the tag <property "name"=...> (required)
+
```
```suggestion
```

jcfr commented on Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:
```diff
@@ -0,0 +1,358 @@
+        break;
+      }
+
+      std::string propertyName = "", propertyValue = "";
+      vtkMRMLAbstractAnnotationPropertyValueProvider::XMLTagAttributes
+          attributes;
+
+      // Get the value for "name" in the tag <property "name"=...> (required)
```
```suggestion
      std::string propertyName = "";
      std::string propertyValue = "";
      vtkMRMLAbstractAnnotationPropertyValueProvider::XMLTagAttributes attributes;

      // Get the value for "name" in the tag <property "name"=...> (required)
```

jcfr commented on Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:
```diff
@@ -0,0 +1,358 @@
+    }
+
+    std::string position = cornerOrEdge->GetAttribute("position");
+
```
```suggestion
```

jcfr commented on Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:
```diff
@@ -0,0 +1,358 @@
+  {
+    if (printWarnings)
+    {
+      vtkErrorWithObjectMacro(textNode,
+                              "<annotations> tag had no nested elements.");
```
```suggestion
      vtkErrorWithObjectMacro(textNode, "<annotations> tag had no nested elements.");
```

jcfr commented on Modules/Loadable/CornerText/qSlicerCornerTextDICOMAnnotationPropertyValueProvider.cxx:
```diff
@@ -0,0 +1,157 @@
+#include <vtkMRMLSliceViewDisplayableManagerFactory.h>
+
+// Slicer includes
+#include <vtkSlicerApplicationLogic.h>
+#include <qSlicerApplication.h>
+#include <qSlicerPythonManager.h>
```
```suggestion
#include <qSlicerApplication.h>
#include <qSlicerPythonManager.h>
#include <vtkSlicerApplicationLogic.h>
```

Make sure to sort includes alphabetically

jcfr commented on Modules/Loadable/CornerText/qSlicerCornerTextDICOMAnnotationPropertyValueProvider.cxx:
```diff
@@ -0,0 +1,157 @@
+#include <vtkAutoInit.h>
+
+// Qt includes
+#include <QSettings>
+#include <QObject>
+#include <QDebug>
+#include <QJsonObject>
+#include <QJsonDocument>
```
```suggestion
#include <QDebug>
#include <QJsonObject>
#include <QJsonDocument>
#include <QObject>
#include <QSettings>
```

jcfr commented on Modules/Loadable/CornerText/qSlicerCornerTextDICOMAnnotationPropertyValueProvider.cxx:
```diff
@@ -0,0 +1,157 @@
+// CornerText includes
+#include "qSlicerCornerTextDICOMAnnotationPropertyValueProvider.h"
+
+// DisplayableManager initialization
+#include <vtkAutoInit.h>
```
```suggestion
```

This should only be includes from  `qSlicerCornerTextModule.cxx` where I would also expect the macro `VTK_MODULE_INIT` to also be used.

jcfr commented on Modules/Loadable/CornerText/qSlicerCornerTextModule.cxx:
```diff
@@ -0,0 +1,228 @@
+//-----------------------------------------------------------------------------
+qSlicerCornerTextModule::~qSlicerCornerTextModule()
+{
+  writeSettings();
```
```suggestion
  this->writeSettings();
```

jcfr commented on Modules/Loadable/CornerText/qSlicerCornerTextModule.cxx:
```diff
@@ -0,0 +1,228 @@
+QStringList qSlicerCornerTextModule::contributors() const
+{
+  QStringList moduleContributors;
+  moduleContributors << QString("John Doe (AnyWare Corp.)");
```
```suggestion
  moduleContributors << QString("Andrew Howe (Kitware)");
  moduleContributors << QString("Jean-Christophe Fillion-Robin (Kitware)");
```

jcfr commented on Modules/Loadable/CornerText/qSlicerCornerTextModule.cxx:
```diff
@@ -0,0 +1,228 @@
+//-----------------------------------------------------------------------------
+QString qSlicerCornerTextModule::helpText() const
+{
+  return "This is a loadable module that can be bundled in an extension";
```
Update this

jcfr commented on Modules/Loadable/CornerText/qSlicerCornerTextModule.cxx:
```diff
@@ -0,0 +1,228 @@
+
+//-----------------------------------------------------------------------------
+void qSlicerCornerTextModule::onLayoutChanged(int layout)
+{
```
```suggestion
{
  qSlicerLayoutManager* layoutManager = qSlicerApplication::application()->layoutManager();
```

jcfr commented on Modules/Loadable/CornerText/qSlicerCornerTextModule.cxx:
```diff
@@ -0,0 +1,228 @@
+            ->sliceView()
+            ->displayableManagerByClassName("vtkMRMLCornerTextDisplayableManager");
+
+    vtkMRMLCornerTextDisplayableManager* cornerTextDisplayableManager;
```
Display error if downcast fails

jcfr commented on Modules/Loadable/CornerText/qSlicerCornerTextModule.cxx:
```diff
@@ -0,0 +1,228 @@
+  // Register default annotation provider
+  vtkNew<vtkMRMLDefaultAnnotationPropertyValueProvider> defultProvider;
+  defultProvider->SetAppLogic(this->appLogic());
+  this->appLogic()->GetCornerTextLogic()->RegisterPropertyValueProvider(
+      "Default", defultProvider);
```
```suggestion
  this->appLogic()->GetCornerTextLogic()->RegisterPropertyValueProvider("Default", defultProvider);
```

jcfr commented on Modules/Loadable/CornerText/qSlicerCornerTextModule.cxx:
```diff
@@ -0,0 +1,228 @@
+  // Register the DICOM property value provider
+#ifdef Slicer_USE_PYTHONQT
+  vtkNew<qSlicerCornerTextDICOMAnnotationPropertyValueProvider> DICOMProvider;
+  this->appLogic()->GetCornerTextLogic()->RegisterPropertyValueProvider(
+      "DICOM", DICOMProvider);
```
```suggestion
  this->appLogic()->GetCornerTextLogic()->RegisterPropertyValueProvider("DICOM", DICOMProvider);
```

jcfr commented on Modules/Loadable/CornerText/qSlicerCornerTextModule.cxx:
```diff
@@ -0,0 +1,228 @@
+  }
+
+  // Register the DICOM property value provider
+#ifdef Slicer_USE_PYTHONQT
```
```suggestion
#ifdef Slicer_BUILD_DICOM_SUPPORT
```

jcfr commented on Modules/Loadable/CornerText/Widgets/qSlicerCornerTextSettingsWidget.h:
```diff
@@ -0,0 +1,76 @@
+// Qt includes
+#include <QWidget>
+
+// Settings Widgets includes
+#include "ctkPimpl.h"
+#include "ctkVTKObject.h"
+#include "qMRMLWidget.h"
+#include "qSlicerCornerTextModuleWidgetsExport.h"
+#include "vtkMRMLScene.h"
+#include "vtkMRMLLayoutLogic.h"
+#include "vtkMRMLCornerTextLogic.h"
```
split as `// MRML includes`, `// CTK includes`, ...

arhowe00 replied:
I did this split, but the organization doesn't always seem completely clear to me

jcfr commented on Libs/MRML/Core/vtkMRMLSliceNode.h:
```diff
@@ -565,6 +566,15 @@ class VTK_MRML_EXPORT vtkMRMLSliceNode : public vtkMRMLAbstractViewNode
+  /// Node reference role for text that is used in the slice view
+  const char* GetCornerAnnotationsTextNodeReferenceRole() { return "cornerText"; }
+  const char* GetCornerAnnotationsTextNodeReferenceMRMLAttributeName() { return "cornerTextRef"; }
+
```
```suggestion
  ///}@

  ///@{
  /// Set/Get text node used to configure the slice view corner annotations
```

jcfr commented on Libs/MRML/Core/vtkMRMLSliceNode.h:
```diff
@@ -18,6 +18,7 @@
 // MRML includes
 #include "vtkMRMLAbstractViewNode.h"
 class vtkMRMLVolumeNode;
+class vtkMRMLTextNode;
```
```suggestion
class vtkMRMLTextNode;
class vtkMRMLVolumeNode;
```

arhowe00 replied:
Hey, while I make these edits, should I always only declare the class if I never touch the implementation? For instance, if I did not access any members of `vtkMRMLAbstractViewNode` in both the .h and .cxx of `vtkMRMLSliceNode`, should it be

```
class vtkMRMLAbstractViewNode;
```
?

jcfr commented on Modules/Scripted/DataProbe/DataProbeLib/DICOMAnnotationPropertyValueProvider.py:
```diff
@@ -0,0 +1,221 @@
+                return LAYER_LABEL
+        return LAYER_ANY
+
+    # A part of the default annotation provider, not abstract. We need a way to
+    # get app logic
+    @staticmethod
+    def GetAppLogic():
+        return slicer.app.applicationLogic()
```
```suggestion
```

jcfr commented on Modules/Scripted/DataProbe/DataProbeLib/DICOMAnnotationPropertyValueProvider.py:
```diff
@@ -0,0 +1,221 @@
+            return ""
+
+        # Get the layers
+        sliceLogic = DICOMAnnotationPropertyValueProvider.GetAppLogic().GetSliceLogic(sliceNode)
```
```suggestion
        sliceLogic = slicer.app.applicationLogic().GetSliceLogic(sliceNode)
```

pieper commented on Modules/Scripted/DataProbe/DataProbe.py:
```diff
@@ -403,7 +402,6 @@ def _createSmall(self):
         """
 
         # this method makes SliceView Annotation
```
Delete stray comment

jcfr commented on Libs/MRML/Core/vtkMRMLAbstractAnnotationPropertyValueProvider.h:
```diff
@@ -0,0 +1,143 @@
+  See COPYRIGHT.txt
+  or http://www.slicer.org/copyright/copyright.txt for details.
+
+  Unless required by applicable law or agreed to in writing, software
```
Test

jcfr commented on Libs/MRML/Core/vtkMRMLAbstractAnnotationPropertyValueProvider.h:
```diff
@@ -0,0 +1,143 @@
+#include <vtkObject.h>
+
+// STD includes
+#include <algorithm>
+#include <cctype>
+#include <string>
+#include <unordered_set>
```
Comment on 4 lines

arhowe00 reviewed (commented):
I rebased all these changes into the previous two commits and force pushed.

jcfr commented:
Following discussions with @lassoan during the [42nd Slicer Project Week](https://projectweek.na-mic.org/PW42_2025_GranCanaria/), we agreed to proceed with the integration once the following points are addressed:

- **Prevent saving the text node in the scene:** While allowing users to update the text interactively is convenient, we anticipate potential changes to the XML attributes or format before the next Slicer release. To avoid serializing an "unstable" structure in the scene, we prefer to disable saving of the text node for now. For reference, [here](https://github.com/Slicer/Slicer/blob/6a7deea0e133ff7d81bfaa9e770b29899fa08848/Libs/MRML/Logic/vtkMRMLApplicationLogic.cxx#L451-L486) is the current XML structure.

- **Update `display-level` to use integers instead of descriptive labels:** The current values (`always`, `sometimes`, `least`) limit extensibility. Switching to an integer-based approach will facilitate the integration of additional levels in the future.

- **Create an issue outlining follow-up work**: We should document the planned improvements for a future pull request, ensuring a clear roadmap for enhancing the annotation system. This should include further refinements to the API, support for additional customization options, and considerations for integrating annotation updates more seamlessly into MRML.

----------

While discussing with @lassoan, we also considered changing the `layer` attribute to `role`, which would provide greater flexibility—for example, allowing annotations to reference other text nodes based on different roles.

@lassoan, do you have any comments or suggestions regarding this change?

### Current Structure:

```xml
<annotations>
  <!-- For corners -->
  <corner position="bottom-left">
    <!-- Generic properties -->
    <property name="VolumeName" layer="label" prefix="L: " display-level="always"/>
    <property name="VolumeName" layer="foreground" prefix="F: " display-level="always"/>
    <property name="VolumeName" layer="background" prefix="B: " display-level="always"/>
  </corner>
  <corner position="top-left">
    <!-- DICOM-specific properties -->
    <property name="PatientName" display-level="sometimes"/>
    <property name="PatientID" prefix="ID: " display-level="always"/>
    <property name="PatientInfo" display-level="sometimes"/>
    <property name="SeriesDate" prefix="B: " layer="background" display-level="sometimes"/>
    <property name="SeriesDate" prefix="F: " layer="foreground" display-level="sometimes"/>
    <property name="SeriesTime" prefix="B: " layer="background" display-level="least"/>
    <property name="SeriesTime" prefix="F: " layer="foreground" display-level="least"/>
    <property name="SeriesDescription" prefix="B: " layer="background" display-level="least"/>
    <property name="SeriesDescription" prefix="F: " layer="foreground" display-level="least"/>
  </corner>
  [...]
</annotations>
```

### Proposed Generalization Using `role`:

```xml
<annotations>
  <!-- For corners -->
  <corner position="bottom-left">
    <!-- Generic properties -->
    <property name="VolumeName" role="label" prefix="L: " display-level="always"/>
    <property name="VolumeName" role="foreground" prefix="F: " display-level="always"/>
    <property name="VolumeName" role="background" prefix="B: " display-level="always"/>
  </corner>
  <corner position="top-left">
    <!-- DICOM-specific properties -->
    <property name="PatientName" display-level="sometimes"/>
    <property name="PatientID" prefix="ID: " display-level="always"/>
    <property name="PatientInfo" display-level="sometimes"/>
    <property name="SeriesDate" prefix="B: " role="background" display-level="sometimes"/>
    <property name="SeriesDate" prefix="F: " role="foreground" display-level="sometimes"/>
    <property name="SeriesTime" prefix="B: " role="background" display-level="least"/>
    <property name="SeriesTime" prefix="F: " role="foreground" display-level="least"/>
    <property name="SeriesDescription" prefix="B: " role="background" display-level="least"/>
    <property name="SeriesDescription" prefix="F: " role="foreground" display-level="least"/>
  </corner>
  [...]
</annotations>
```

arhowe00 committed:
ENH: Improve CornerText usability

- Prevent saving to scene

- Hide from editor by default

- Implement CornerText restoreDefaults() button

- Cast bool to int for DataProbe config consistency.

  The previous implementation of `writeSettings()` in
  `Modules/Loadable/CornerText/qSlicerCornerTextModule.cxx` wrote bools
  to the config as true/false as opposed to 1/0. This was incompatible
  with how DataProbe wrote config and is generally less portable, since
  true/false are C++ literals.

- Change function and style of display levels

  Within the XML, the annotator will provide numbers describing the
  display level, as opposed to "always", "sometimes", and "least".
  Furthermore, the "Display Level" now functions as a cutoff, where a
  smaller number indicates that less information will be shown on the
  screen.  This translates to the display level being a cutoff, where
  everything with a display level greater than the cutoff will be
  excluded, and anything with a display level at or below the cutoff is
  included in the display.

- Adjust defaults for CornerText

- Change "layer" to "role" in CornerText

  Now, the XML specifies "role" instead of "layer", which is more
  generalizable.

Co-authored-by: Jean-Christophe Fillion-Robin <jchris.fillionr@kitware.com>

arhowe00 commented:
Thanks @jcfr, I implemented these changes into another commit which I've pushed.

## Summary

The existing refactoring will allow any module to dynamically modify annotations in a slice view. The implementation could be more lean with a few changes, but there are tradeoffs in those changes:

1. MRML attributes could be directly added and referenced when determining what is displayed in a corner, which would reduce the processing that has to occur on part of an annotation provider.  This could potentially constrain an implementor that wants a group of computations to decide what is displayed on the screen.
2. Regarding the XML, it may provide a good middle ground between "our-module-specific-jargon" and a readable way for users to modify the displayed text.  However, removing the XML would make the API less reliant on a custom, albeit lightweight, syntax, allowing complete programmatic display of strings on the slice view.  Such a programmatic view might be more modular, where clashing modules which all want to contribute their own annotations to the slice view could have ranking or floating values associated with the annotations, with annotations displayed in the order of their rank.  Then, annotations from anywhere can be composed and rendered, allowing a more complete answer to all possible needs of users (clinicians, researchers, industry users).

Furthermore, there are a few more things of note:

1. We deferred the decision on removing the "Background DICOM annotations persistence" button.
2. We should keep in mind dependencies who use the class `SliceAnnotations`.  In addition to providing the new flexibility, registration APIs, and refactoring into C++, this PR also implements the annotations features provided by `SliceViewAnnotations.py`, but is not a drop-in-substitute for dependencies of the SliceAnnotations Python class.
3. I like the idea of switching "layer" to "role", but note that both the current implementors ("DICOM" and "Default") use the concept of "role" to mean layer. As you can see, for instance, both providers still use the functions `vtkMRMLSliceLogic::GetForegroundLayer()`, as the "role" which `vtkMRMLDefaultAnnotationPropertyValueProvider.cxx` is interested in.  Other implementors can do whatever they want with "role", but they could also have just made a custom property and parsed it themselves in their provider.
4. Note how the new behavior of "Display Level" is that a lower number will show less information on the screen.  I have tried to make the implementation of "Display Level" as intuitively as I possibly could in the code.
5. We could give text nodes specific to the annotations module a custom property and then select and edit them from within the CornerText module -- this is a WIP.

jcfr commented:
@arhowe00  Could you have a look at the following failure [reported](https://github.com/Slicer/Slicer/actions/runs/13655531787/job/38173691006) in the CI ?

```
/usr/src/Slicer/Libs/MRML/Logic/vtkMRMLCornerTextLogic.h:118:14: error: ‘unordered_map’ in namespace ‘std’ does not name a template type
   const std::unordered_map<std::string, TextLocation> positionMap =
              ^~~~~~~~~~~~~
/usr/src/Slicer/Libs/MRML/Logic/vtkMRMLCornerTextLogic.h:138:8: error: ‘unordered_map’ in namespace ‘std’ does not name a template type
   std::unordered_map<
        ^~~~~~~~~~~~~
/usr/src/Slicer/Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx: In member function ‘virtual void vtkMRMLCornerTextLogic::PrintSelf(std::ostream&, vtkIndent)’:
/usr/src/Slicer/Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:60:37: error: ‘class vtkMRMLCornerTextLogic’ has no member named ‘registeredProviders’; did you mean ‘RegisterNodes’?
   for (const auto& provider : this->registeredProviders)
                                     ^~~~~~~~~~~~~~~~~~~
                                     RegisterNodes
/usr/src/Slicer/Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx: In member function ‘bool vtkMRMLCornerTextLogic::RegisterPropertyValueProvider(const string&, vtkMRMLAbstractAnnotationPropertyValueProvider*)’:
/usr/src/Slicer/Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:132:7: error: ‘registeredProviders’ was not declared in this scope
   if (registeredProviders.find(pluginName) == registeredProviders.end())
       ^~~~~~~~~~~~~~~~~~~
/usr/src/Slicer/Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:132:7: note: suggested alternative: ‘RegisterNodes’
   if (registeredProviders.find(pluginName) == registeredProviders.end())
       ^~~~~~~~~~~~~~~~~~~
       RegisterNodes
/usr/src/Slicer/Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx: In member function ‘std::array<std::basic_string<char>, 8> vtkMRMLCornerTextLogic::GenerateAnnotations(vtkMRMLSliceNode*, vtkMRMLTextNode*, bool)’:
/usr/src/Slicer/Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:260:16: error: ‘class vtkMRMLCornerTextLogic’ has no member named ‘positionMap’
     if (!this->positionMap.count(position))
                ^~~~~~~~~~~
/usr/src/Slicer/Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:272:35: error: ‘class vtkMRMLCornerTextLogic’ has no member named ‘positionMap’
     TextLocation location = this->positionMap.at(position);
                                   ^~~~~~~~~~~
/usr/src/Slicer/Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:335:44: error: ‘registeredProviders’ was not declared in this scope
       for (const auto [plugin, provider] : registeredProviders)
                                            ^~~~~~~~~~~~~~~~~~~
/usr/src/Slicer/Libs/MRML/Logic/vtkMRMLCornerTextLogic.cxx:335:44: note: suggested alternative: ‘RegisterNodes’
       for (const auto [plugin, provider] : registeredProviders)
                                            ^~~~~~~~~~~~~~~~~~~
                                            RegisterNodes
```

jcfr commented:
Summary of changes:
- Fix seg fault on startup ensuring pointers are all initialized
- Fix memory leak updating `vtkMRMLCornerTextLogic::ParseTextNode` to accept `vtkXMLDataParser*` as argument and remove calls to `Register`
- Fix memory leak switch `vtkMRMLDefaultAnnotationPropertyValueProvider:AppLogic` to use `vtkWeakPointer`
- Fix formatting of pointers and references (prefer `vtkObject*` and `std::string&` instead of `vtkObject *` and `std::string &`)
- Other minor tweaks

jcfr commented:
Before integrating, we should dispatch the changes currently associated with the usability commit into their respective ones.
